### ISSUE TITLE
The same driver for program and set device cannot be used

### ISSUE THREAD

**Issue by Samarthjainabout** [lekez2005/opennvram-reram#8]
HI, 
As per my previous publications, we need to make separate drivers for forming the device (~3v) and program device(~1v)
Can u explain if you have made separate drivers for these?
I could not find them in the project file.

**Comment by lekez2005** [comment]
Can you point me to the publication you're referring to? 

I separated out the wordline and write driver voltages to make it possible to configure the voltage used. If you look at the SRAM pins, you'll see `vdd_write_bl vdd_write_br vdd_wordline` pins. I agree that having completely separate drivers might potentially be a better long-term solution. But also, forming needs to be done just once, so maybe one driver could be fine?

**Comment by Samarthjainabout** [comment]
This is an Old publication, my latest one is still under journal review. I can share that details only on zoom if you are interested,
So your vdd_wordline pin is for forming? and write is for set,reset? 
[SSDM2022_Li Sifan_v1.pdf](https://github.com/lekez2005/opennvram-reram/files/11207486/SSDM2022_Li.Sifan_v1.pdf)

**Comment by lekez2005** [comment]
Yes, something like that. More precisely, the testing plan was to start out with the voltages published [here](https://sky130-fd-pr-reram.readthedocs.io/en/latest/technology_specifications.html#seting-reseting) which are more specific the Skywater process.

I'm happy to schedule a call at some point to discuss further.

**Comment by Samarthjainabout** [comment]
> I separated out the wordline and write driver voltages to make it possible to configure the voltage used. If you look at the SRAM pins, you'll see `vdd_write_bl vdd_write_br vdd_wordline` pins. I agree that having completely separate drivers might potentially be a better long-term solution. But also, forming needs to be done just once, so maybe one driver could be fine?

which spice file is this defined in?

**Comment by lekez2005** [comment]
You can take a look at the full spice file at for example $CARAVEL_WORKSPACE/netgen/r_16_w_16.spice. Write driver spice nelist is at https://github.com/lekez2005/opennvram-reram/blob/main/technology/sky130/sp_lib/write_driver_mux_buffer.sp

**Comment by Samarthjainabout** [comment]
> Yes, something like that. More precisely, the testing plan was to start out with the voltages published [here](https://sky130-fd-pr-reram.readthedocs.io/en/latest/technology_specifications.html#seting-reseting) which are more specific the Skywater process.
> 
> I'm happy to schedule a call at some point to discuss further.

As per the device details in link, it is mentioned the forming needs to happen in a Ramp mapper with a voltage step of 0.1V, but the driver made in the chip can only give binary pulses.I doubt that will give a good yield.
Generally when I made my own RRAM, forming device is most imp and fragile process and should be done in a controlled manner.
How did you account for it in driver I am little curious?
![image](https://user-images.githubusercontent.com/30805083/231972179-57c471ed-d14a-443f-a9fa-823c074a7e13.png)

**Comment by lekez2005** [comment]
For "forming", I think the "ramp" can be done at the board-level by externally "ramping" `vdd_write_bl vdd_write_br vdd_wordline`. The ramping can be easily done using appropriate programmable DC sources. 

Is the ramp also required for all sets/resets? Or maybe the appropriate set/reset voltages can be determined once and fixed for subsequent set/resets?

**Comment by Samarthjainabout** [comment]
I have got data showing ramping for set/reset in general and specific my rram gives superior performance in terms of state variability
![image](https://user-images.githubusercontent.com/30805083/232377547-6b6b3358-b000-4da4-a738-15a4e71e693b.png)
![image](https://user-images.githubusercontent.com/30805083/232377568-96fdb6de-a367-4021-b8e0-5f8cc72b0e74.png)

**Comment by lekez2005** [comment]
Interesting, is the cycle-to-cycle variability strongly influenced by the ramp rate?

**Comment by Samarthjainabout** [comment]
yes,if pulse is not ramped, the cycle to cycles variability is bad. I have tried many different schemes for my own work and that was my conclusion. you can google literature, its same for most CB RAM
